### PR TITLE
Refactor Values, Slice, and Kvslice in D/O/Array,Hash

### LINKED ISSUE #268: Data::Object::Hash::Func::Slice - confusing name
I think that [Data::Object::Hash::slice](https://metacpan.org/pod/Data::Object::Hash#slice) may be a confusing name. In perl, there are two types of hash slices:

```
my %hash = (
    a => 'b',
    c => 'd',
);
```

- [value slice](https://perlmaven.com/hash-slice)

```
my ($a, $c) = @hash{qw/a c/};

# $a holds 'b'
# $c holds 'd'
```

- [keyvalue slice](https://www.effectiveperlprogramming.com/2014/07/perl-5-20-introduces-keyvalue-slices/)
```
my @keyvalue_slice = %hash{qw/a c/};

# @keyvalue_slice holds 'a', 'b', 'c', 'd'
```

The current `slice` method is almost like perl 5.20's keyvalue slice, but it wraps it in a hashref.

### PR DESCRIPTION
**Pull request by iamalnewkirk**
Closes #268, Cancels #297, On-Behalf-Of @srchulo

### CODE REVIEW

**Comment by iamalnewkirk** [comment]
@srchulo See test failures, https://github.com/iamalnewkirk/do/pull/304/checks?check_run_id=229388556#step:11:1496

**Comment by srchulo** [comment]
@iamalnewkirk it looks like I'm unable to push to this branch.

**Comment by iamalnewkirk** [comment]
@srchulo You should have access now. This project uses a rebase merge strategy. Please try to avoid merge commits when pushing branches.

File: lib/Data/Object/Array/Func/Kvslice.pm [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+has arg1 => (
+  is => 'ro',
+  isa => 'ArrayLike',
+  req => 1
+);
+
+has args => (
+  is => 'ro',
+  isa => 'ArrayRef[Any]',
```
Comment (iamalnewkirk) [inline lib/Data/Object/Array/Func/Kvslice.pm]: This should probably be `ArrayRef[Int]`.

**Comment by srchulo** [comment]
@iamalnewkirk thanks! The push worked :)

File: lib/Data/Object/Hash/Func/Kvslice.pm [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+has arg1 => (
+  is => 'ro',
+  isa => 'HashLike',
+  req => 1,
+);
+
+has args => (
+  is => 'ro',
+  isa => 'ArrayRef[Any]',
```
Comment (iamalnewkirk) [inline lib/Data/Object/Hash/Func/Kvslice.pm]: This should probably be `ArrayRef[Str]`.

File: t/0.90/can/Data_Object_Array_kvslice.t [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+=description
+
+The kvslice method returns an array reference containing the elements in the
+array at the index(es) specified in the arguments. This method returns a
+L<Data::Object::Hash> object.
+
+=signature
+
+kvslice(Any $arg1) : Any
```
Comment (iamalnewkirk) [inline t/0.90/can/Data_Object_Array_kvslice.t]: This should be `kvslice(Any @args) : Any`.

File: t/0.90/can/Data_Object_Array_kvslice.t [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+=usage
+
+  # given [1..5]
+
+  $array->kvslice(2,4); # {2=>3, 4=>5}
+
+=description
+
+The kvslice method returns an array reference containing the elements in the
```
Comment (iamalnewkirk) [inline t/0.90/can/Data_Object_Array_kvslice.t]: I think `returns an array reference` should be `returns a hash reference`.

File: t/0.90/can/Data_Object_Hash_kvslice.t [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+=description
+
+The kvslice method returns a hash reference containing the elements in the hash
+at the key(s) specified in the arguments. This method returns a
+L<Data::Object::Hash> object.
+
+=signature
+
+kvslice(Any $arg1) : HashObject
```
Comment (iamalnewkirk) [inline t/0.90/can/Data_Object_Hash_kvslice.t]: This should probably be `kvslice(Str @args) : HashObject`.

File: t/0.90/can/Data_Object_Hash_slice.t [hunk 6]
```diff
@@ -15,17 +15,17 @@ slice
...
-at the key(s) specified in the arguments. This method returns a
-L<Data::Object::Hash> object.
+The slice method returns an array reference of the values that correspond to
+the key(s) specified in the arguments. This method returns a
+L<Data::Object::Array> object.
 
 =signature
 
-slice(Any $arg1) : Any
+slice(Str $arg1) : ArrayObject
```
Comment (iamalnewkirk) [inline t/0.90/can/Data_Object_Hash_slice.t]: Should be `slice(Str @args) : ArrayObject`.

File: t/0.90/use/Data_Object_Array_Func_Kvslice.t [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  $func->execute;
+
+=inherits
+
+Data::Object::Array::Func
+
+=attributes
+
+arg1(ArrayLike, req, ro)
+args(ArrayRef[Any], req, ro)
```
Comment (iamalnewkirk) [inline t/0.90/use/Data_Object_Array_Func_Kvslice.t]: Should be `args(ArrayRef[Str], req, ro)`.

File: t/0.90/use/Data_Object_Hash_Func_Kvslice.t [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+  $func->execute;
+
+=inherits
+
+Data::Object::Hash::Func
+
+=attributes
+
+arg1(Object, req, ro)
+args(ArrayRef[Any], req, ro)
```
Comment (iamalnewkirk) [inline t/0.90/use/Data_Object_Hash_Func_Kvslice.t]: Should be `args(ArrayRef[Str], req, ro)`.

File: t/0.90/use/Data_Object_Hash_Func_Slice.t [hunk 9]
```diff
@@ -30,7 +30,7 @@ Data::Object::Hash::Func
 =attributes
 
 arg1(Object, req, ro)
-args(ArrayRef[Any], req, ro)
+args(ArrayRef[StringLike], opt, ro)
```
Comment (iamalnewkirk) [inline t/0.90/use/Data_Object_Hash_Func_Slice.t]: 👍

File: lib/Data/Object/Array/Func/Kvslice.pm (hunk 1 above)
Comment (srchulo) [reply lib/Data/Object/Array/Func/Kvslice.pm]: For Array should it be `ArrayRef[NumberLike]` instead?

Whichever we use, I'll make the same change in `Array::Func::Slice`, as it also uses `Any`.

Comment (iamalnewkirk) [reply lib/Data/Object/Array/Func/Kvslice.pm]: In Perl, and PHP (fwiw), `$array['this'] = 'that'` is legal. So `ArrayRef[StringLike]` is safer, even though that usage should be discouraged.

File: lib/Data/Object/Hash/Func/Kvslice.pm (hunk 2 above)
Comment (iamalnewkirk) [reply lib/Data/Object/Hash/Func/Kvslice.pm]: In Perl, and PHP (fwiw), `$array['this'] = 'that'` is legal. So `ArrayRef[StringLike]` is safer, even though that usage should be discouraged.

File: t/0.90/can/Data_Object_Array_kvslice.t (hunk 4 above)
Comment (srchulo) [reply t/0.90/can/Data_Object_Array_kvslice.t]: Good catch!

File: lib/Data/Object/Array/Func/Kvslice.pm (hunk 1 above)
Comment (srchulo) [reply lib/Data/Object/Array/Func/Kvslice.pm]: yuck! :)

**Pull request merged**